File: jaeseuk/kokotlin/src/main/kotlin/programmers/level1/SortBoxers.kt
package programmers.level1

class SortBoxers {
    fun solution(weights: IntArray, head2head: Array<String>): IntArray {
        val answer = arrayListOf<Array<Any>>()

        for (i in 0..weights.lastIndex) {
            val games = head2head[i].length - head2head[i].count { it == 'N' }
            val win = if (games > 0)
                head2head[i].count { it == 'W' }.toDouble() / games else 0
            val cnt = head2head[i].withIndex().count {
                it.value == 'W' && weights[i] < weights[it.index]
            }

            answer.add(arrayOf(win, cnt, weights[i], i + 1))
        }

        return answer
            .sortedWith(
                compareBy(
                    { -it[0].toString().toDouble() },
                    { -it[1].toString().toInt() },
                    { -it[2].toString().toInt() },
                    { it[3].toString().toInt() })
            )
            .map { it[3].toString().toInt() }.toIntArray()
    }
}

File: jaeseuk/pypython/Programmers/level1/sort_boxers.py
def solution(weights, head2head):
    answer = []

    for idx, player in enumerate(zip(weights, head2head)):
        game = len(player[1]) - player[1].count("N")
        win = player[1].count("W") / game if game > 0 else 0
        cnt = 0

        for i in range(len(player[1])):
            if idx != i and player[1][i] == "W" and player[0] < weights[i]:
                cnt += 1

        answer.append((idx + 1, win, cnt, player[0]))

    answer.sort(key=lambda x: (-x[1], -x[2], -x[3], x[0]))
    return list(map(lambda x: x[0], answer))
